src/Test.UnitTests.Sarif.WorkItems/SarifWorkItemFilerTests.cs

```diff
@@ -3,17 +3,165 @@
 
 using System;
 using System.Collections.Generic;
+using System.IO;
+using System.Threading;
 using System.Threading.Tasks;
 using FluentAssertions;
+using Microsoft.TeamFoundation.WorkItemTracking.WebApi.Models;
+using Microsoft.VisualStudio.Services.WebApi;
+using Microsoft.VisualStudio.Services.WebApi.Patch.Json;
 using Microsoft.WorkItems;
 using Moq;
+using Newtonsoft.Json;
 using Xunit;
 
 namespace Microsoft.CodeAnalysis.Sarif.WorkItems
 {
     public class SarifWorkItemFilerTests
     {
-        private static readonly Uri s_testUri = new Uri("https://github.com/Microsoft/sarif-sdk");
+        [Fact]
+        public void WorkItemFiler_AzureDevOpsClientReceivesExpectedCalls()
+        {
+            // IMPORTANT. This is a single end-to-end test that demonstrates we have
+            // the capability of mocking/injecting mocks into the end-to-end work
+            // item filing scenario. This example works for the ADO filer only. Once
+            // reviewed/finalized, we would look to provide an equivalent implementation
+            // for the GitHub filer. The code below isn't intended to reflect the 
+            // actual factoring we can expect when we use this machinery to cover
+            // all expected future cases. i.e., after completing this first review,
+            // we should look at more thoughtful factoring/helpers to capture 
+            // testing for specific scenarios.
+
+            // ONE. Create test data that the low-level ADO client mocks
+            //      will flow back to the SARIF work item filer. 
+            var attachmentReference = new AttachmentReference()
+            {
+                Id = Guid.NewGuid(),
+                Url = Guid.NewGuid().ToString()
+            };
+
+            var workItem = new WorkItem
+            {
+                Id = DateTime.UtcNow.Millisecond,
+                Links = new ReferenceLinks()
+            };
+
+            // The fake URI to the filed item that we'll expect the filer to receive
+            string htmlUri = "https://example.com/" + Guid.NewGuid().ToString();
+            workItem.Links.AddLink("html", htmlUri);
+
+            // TWO. Define variables to capture whether we enter all expected ADO client methods.
+            bool connectCalled = false,
+                 createWorkItemCalled = false,
+                 createAttachmenCalled = false;                 
```

Comment:
>false [](start = 41, length = 5)

Moq has a facility for verifying calls: [https://stackoverflow.com/questions/9136674/verify-a-method-call-using-moq#9136721](https://stackoverflow.com/questions/9136674/verify-a-method-call-using-moq#9136721)

Reply:
Noted, thanks.

---
In reply to: [389390695](https://github.com/microsoft/sarif-sdk/pull/1800#discussion_r389390695) [](ancestors = 389390695)